ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_parser.rs

```diff
@@ -0,0 +1,309 @@
+    pub async fn process_next_batch(
+        &mut self,
+        db: DbReaderWriter,
+    ) -> Vec<Result<EndVersion, Status>> {
+        let mut tasks = vec![];
+        let (batches, total_txns_to_process) = self.get_batches().await;
+        let db_writer = db.writer.clone();
+
+        for batch in batches {
+            let start_time = std::time::Instant::now();
+            let db_writer = db_writer.clone();
+            let context = self.context.clone();
+            let ledger_version = self.highest_known_version;
+            let task = tokio::spawn(async move {
+                let raw_txns =
+                    Self::fetch_raw_txns_with_retries(context.clone(), ledger_version, batch).await;
+                Self::parse_table_info(context.clone(), raw_txns.clone(), db_writer.clone())
+                    .expect("[Table Info] Failed to parse table info");
+
+                DURATION_IN_SECS
+                    .with_label_values(&[
+                        SERVICE_TYPE,
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_step(),
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_label(),
+                    ])
+                    .set(start_time.elapsed().as_secs_f64());
+                TOTAL_SIZE_IN_BYTES
+                    .with_label_values(&[
+                        SERVICE_TYPE,
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_step(),
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_label(),
+                    ])
+                    .set(bcs::serialized_size(&raw_txns).unwrap() as i64);
+
+                Ok(raw_txns.last().unwrap().version)
+            });
+            tasks.push(task);
+        }
+        match futures::future::try_join_all(tasks).await {
+            Ok(res) => {
+                // After all threads finish processing, clean up pending on items across threads
+                db_writer
+                    .clone()
+                    .cleanup_pending_on_items()
+                    .expect("[Table Info] Failed to clean up the pending on items");
+                let db_reader = db.reader.clone();
+                let mut retried: u64 = 0;
+
+                // If pending on items are not empty, meaning the current loop hasn't fully parsed all table infos
+                // due to the nature of multithreading where instructions used to parse table info might come later,
+                // retry sequentially to ensure parsing is complete
+                //
+                // Risk of this sequential approach is that it could be slow when the txns to process contain extremely
+                // nested table items, but the risk is bounded by the the configuration of the number of txns to process and number of threads
+                while !db_reader
+                    .clone()
+                    .is_indexer_async_v2_pending_on_empty()
+                    .unwrap()
```

Comment:
If you'll sequentially parse all of them this should be done one time. You shouldn't need a while loop. In fact you should crash if it's still empty after one run.